### ISSUE TITLE
sd-event Sometimes doesn't process the bus in a timely manner

### ISSUE THREAD

**Issue by wak-google** [systemd/systemd#22046]
**systemd version the issue has been seen with**
> 249.6

**Used distribution**
 > Yocto openembedded

**Linux kernel version used** (`uname -a`)
 > `5.15.10`

**CPU architecture issue was seen on**
 > arm

**Expected behaviour you didn't see**
 > The daemon processing all of the bus events and sharing the CPU about as much as the producer.

**Unexpected behaviour you saw**
 > The daemon consuming bus matches using almost 0 CPU while the signal producer is using ~20% CPU purely generating signals on the bus. The dbus-broker ends up with a backlog of signals and the consumer also ends up inflating memory with incoming events.

**Steps to reproduce the problem**
 > Start the signal producer before starting the consumer. This is important as it appears these failures only occur if the consumer (using sd_bus_attach_event and sd_event_loop never calling sd_bus_process on its own) is being started while encountering a heavy load of incoming events. I assume this has something to do with setting up the match but never calling sd_bus_process before entering the event loop.

**Additional program output to the terminal or log subsystem illustrating the issue**

```text
…
```

**Comment by wak-google** [comment]
Working on making a better reproducer as getting our postcode reporting daemons running outside our environment is kinda tedious even though they are open source.
https://github.com/openbmc/phosphor-host-postd
https://github.com/openbmc/phosphor-watchdog

**Comment by wak-google** [comment]
Also, you can patch the sd-bus event handlers to work around the issue, even though I'm pretty sure this wouldn't be an accepted fix.
```
diff --git a/src/libsystemd/sd-bus/sd-bus.c b/src/libsystemd/sd-bus/sd-bus.c
index 9e1d29cc1d..ad424c742f 100644
--- a/src/libsystemd/sd-bus/sd-bus.c
+++ b/src/libsystemd/sd-bus/sd-bus.c
@@ -3592,15 +3592,17 @@ bool bus_pid_changed(sd_bus *bus) {
         return bus->original_pid != getpid_cached();
 }
 
-static int io_callback(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+static bus_process_callback(void *userdata) {
         sd_bus *bus = userdata;
         int r;
 
         assert(bus);
 
-        /* Note that this is called both on input_fd, output_fd as well as inotify_fd events */
+        do
+        {
+                r = sd_bus_process(bus, NULL);
+        } while (r > 0);
 
-        r = sd_bus_process(bus, NULL);
         if (r < 0) {
                 log_debug_errno(r, "Processing of bus failed, closing down: %m");
                 bus_enter_closing(bus);
@@ -3609,19 +3611,13 @@ static int io_callback(sd_event_source *s, int fd, uint32_t revents, void *userd
         return 1;
 }
 
-static int time_callback(sd_event_source *s, uint64_t usec, void *userdata) {
-        sd_bus *bus = userdata;
-        int r;
-
-        assert(bus);
-
-        r = sd_bus_process(bus, NULL);
-        if (r < 0) {
-                log_debug_errno(r, "Processing of bus failed, closing down: %m");
-                bus_enter_closing(bus);
-        }
+static int io_callback(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        /* Note that this is called both on input_fd, output_fd as well as inotify_fd events */
+        return bus_process_callback(userdata);
+}
 
-        return 1;
+static int time_callback(sd_event_source *s, uint64_t usec, void *userdata) {
+        return bus_process_callback(userdata);
 }
 
 static int prepare_callback(sd_event_source *s, void *userdata) {
```

**Comment by wltu** [comment]
Something to note from the system that had this issue.

I am seeing that the sdevent + signal mater + ipmi calls will result in the issue. The ipmi watchdog command will call the dbus for watchdog and probably contributed to the issue. Disabling the ipmi handler will result in us not hitting the issue.

Just something that I noticed, not sure if it matters.

**Comment by poettering** [comment]
i am not sure i follow? what's the precise issue here? note that D-Bus is not usable to "stream" data, as sending too many messages in a time window will cause the broker to throttle you and eventually kick you off the bus. D-bus is purely suitable for control data, not for payload data hence.

I udnerstand your test program floods signals into the bus and you are surprised these don#t arrive? That's not surprising, and kinda a feature, not a bug: participants on the bus should not be able to flood the bus...

What am I missing?

**Comment by wak-google** [comment]
No, the problem is that d-bus broker and the process receiving the signals end up consuming unbounded amounts of memory. Presumably this is because the event loop is not actually waking up for signals that need to be processed by sd_bus_process. The above patch changes the behavior of the event processing loop to be edge style instead of regular poll style. By consuming all of the outstanding events that need processing we never see a resource leak. If we only consume one event at a time and go back to epoll(), our consumer eventually stops being notified and bus resources pile up.

I would be perfectly fine if the bus silently started dropping too many signals and the client application didn't start leaking bus resources.

**Comment by wak-google** [comment]
It appears that the number of events on the file descriptor being epoll()'d is fewer than the number of sd_bus_process() calls needed.

**Comment by poettering** [comment]
> No, the problem is that d-bus broker and the process receiving the signals end up consuming unbounded amounts of memory.

What makes you think it's unbounded?

Both dbus-broker and sd-bus will maintain in-memory queues of queued-but-not-written messages for each connection. But these are bounded, there's a limit on it. Once the limit is hit, the thing fails.

> Presumably this is because the event loop is not actually waking up for signals that need to be processed by sd_bus_process. The above patch changes the behavior of the event processing loop to be edge style instead of regular poll style. By consuming all of the outstanding events that need processing we never see a resource leak. If we only consume one event at a time and go back to epoll(), our consumer eventually stops being notified and bus resources pile up.

sd-bus is an asynchronous library: it will not block on IO, but only do IO when the event loop signals us the socket is writable. We'll queue messages locally as mentioned if the socket is momentarily not writable. This means that if the receiving side is not processing messages quickly enough we'll queue locally in memory. But this all has limits, once the limit is hit, we'll error out.

> I would be perfectly fine if the bus silently started dropping too many signals and the client application didn't start leaking bus resources.

The (hard-coded) limits on queued messages per connection in sd-bus are 384K right now. Not sure what limits dbus-broker is enforcing.

**Comment by wltu** [comment]
we have hit cases where the sdevent pile up so much that the OOM-killer had to come in and clean it out.

Needed the following to resolve the issue.

https://github.com/openbmc/phosphor-watchdog/commit/6f80f80e02f220d3d13a0a6271fa1afd54dc706d

Not sure if this help at all.

**Comment by wak-google** [comment]
> > No, the problem is that d-bus broker and the process receiving the signals end up consuming unbounded amounts of memory.
> 
> What makes you think it's unbounded?
> 
> Both dbus-broker and sd-bus will maintain in-memory queues of queued-but-not-written messages for each connection. But these are bounded, there's a limit on it. Once the limit is hit, the thing fails.
> 
Right I completely understand that what you wrote is the behavior we should be observing, but it doesn't seem like those limits are respected as dbus-broker and the service receiving the match data end up consuming 100's of MB of memory until OOM killer is activated. If we just modify the code to drain all of the dbus events instead of 1 per invocation of an io-event or timer expiration, we never run into this scenario. It would appear that we have some kind of lack of triggering from the kernel (similar to edge triggering and not draining all outstanding IOs). But at a glance it looks like the io callback is not set to be edge triggered.

> > Presumably this is because the event loop is not actually waking up for signals that need to be processed by sd_bus_process. The above patch changes the behavior of the event processing loop to be edge style instead of regular poll style. By consuming all of the outstanding events that need processing we never see a resource leak. If we only consume one event at a time and go back to epoll(), our consumer eventually stops being notified and bus resources pile up.
> 
> sd-bus is an asynchronous library: it will not block on IO, but only do IO when the event loop signals us the socket is writable. We'll queue messages locally as mentioned if the socket is momentarily not writable. This means that if the receiving side is not processing messages quickly enough we'll queue locally in memory. But this all has limits, once the limit is hit, we'll error out.
> 
Yeah, I would expect an error to but we never get such a thing. It seems like the queues on both sides are ultimately unbounded and forcefully processing them fixes the behavior.
> > I would be perfectly fine if the bus silently started dropping too many signals and the client application didn't start leaking bus resources.
> 
I don't think it needs to, the client side never processes the messages and just inflates in memory size. It has almost no CPU scheduling time when profiled.
> The (hard-coded) limits on queued messages per connection in sd-bus are 384K right now. Not sure what limits dbus-broker is enforcing.
Yeah, but we go way beyond this limit in our testing so something unexpected is happening here.
